src/txn/txn_ckpt.c

```diff
@@ -189,10 +211,19 @@ __checkpoint_apply(WT_SESSION_IMPL *session, const char *cfg[],
 			continue;
 		WT_WITH_DHANDLE(session, session->ckpt_handle[i],
 		    ret = (*op)(session, cfg));
-		WT_RET(ret);
+		WT_ERR(ret);
 	}
 
-	return (0);
+err:
+	/*
+	 * If we have an error somewhere in processing the handles, then
+	 * we need to mark earlier trees dirty.  Recursively call the apply
+	 * function to reset on error.
+	 */
+	if (ret != 0)
+		(void)__checkpoint_apply(session, cfg, __checkpoint_err_reset);
```

Comment:
I'm concerned that this is recursive, I can see that ``__checkpoint_err_reset`` doesn't currently return an error, so we wouldn't end up in an infinite recursion - just the possibility of that in the future makes me nervous though. How would you feel about moving the call to reset into the error handler of ``__txn_checkpoint``? Have we already given up the list of handles by then?